### DISCUSSION TITLE
Add Methods in Module To Stream Multi Results

### DISCUSSION THREAD

**Discussion by jrpetersjr** [henrikfroehling/Trakt.NET#415] [Ideas]
GetMultipleCommentsAsync, GetMultipleEpisodesAsync, GetMultipleMoviesAsync, GetMultiplePersonsAsync, GetMultipleSeasonsAsync, GetMultipleShowsAsync, and GetMultiplePersonalListsAsync all have Task.WhenAll call in them for multiple API calls. The problem comes when the trakt api is slow or trying to retrieve a large amount of values, or the HTTPClient has a small amount of of Max Connections Per Server (.Net Framework defaults to 2). For instance I have 35+ lists. If I want to get all of my lists and 1 of those lists takes 5 seconds to pull from trakt because their API slows down at that point it will take 5 seconds before seeing any data and it's basically the same thing for if my "MaxConnectionsPerServer" is smaller than the amount of items that I'm trying to pull.

I'm proposing that an extra method is added for each of these that streams the values back as they finish. This allows processing while we wait for other API calls to returns. The name of the methods could be like this "GetThingStreamAsync" and the method signatures would almost be the same, but instead of return Task<IEnumerable<X>> they would return IAsyncEnumerable<X>. This would allow the library to yield return values as they come back from the API instead of waiting for them all to finish. 

To consume an IAsyncEnumerable you can do await foreach(var thing in thingThatsIAsyncEnumerable) and an extension method for IAsyncEnumerable can be created so that you can make it into a list so that you can pass it around as a list later if needed. It could be called ToListAsync(). 

IAsyncEnumerable is available in a lot of newer versions of .Net, but isn't available OOB for .NetStandard2.0 (it's available for 2.1 and up). In order to do this [Microsoft Provided Nuget Package](https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces/) would need to be used. This could give better performance and compatibility with newer features introduced in newer version of .Net. 

I can make these updates myself, but I wanted to reach out to get your feelings on the idea before I made an issue and then started working on it just like the guidelines say. If you need an example let me know.

TIA

**Comment by henrikfroehling** [comment]
Good catch.

The `GetMultiple...Async` methods were never meant for large request lists, but just as convenience methods for like a small amount of similar requests.
Maybe I should have made that more clear in the comments of those methods.

The existing methods should stay as they are. But, I think, they can be marked as deprecated and then removed in a later version.

I also don't have a problem with adding a new dependency, if it serves a purpose in this case.

The following steps should then be implemented.
1. Add new dependency for `IAsyncEnumerable`
2. Add new method overload with the name `Get...StreamAsync` => for `GetMultipleCommentsAsync`, this would look like this: `public async IAsyncEnumerable<TraktResponse<ITraktComment>> GetCommentsStreamAsync(uint[] commentIds, CancellationToken cancellationToken = default)`
    - This should be done for all `GetMultiple...Async` methods in all modules
3. Deprecating the existing methods

This can also be implemented for the upcoming version 1.4.0 on which I want to start in the next weeks.

**Comment by henrikfroehling** [comment]
I created the issue #416 for this.

**Reply by jrpetersjr** [reply]
I can have this done in a couple of days. Is there a specific branch that I should be developing against or just fork develop?

**Reply by henrikfroehling** [reply]
For version 1.4.0 there is already the branch `release-1.4.0`, which should be the target for this.

**Reply by jrpetersjr** [reply]
Thanks. 

There's one thing that I missed adding to the original comment, but the CancellationToken in the signature would need to be changed to "[EnumeratorCancellation]  CancellationToken cancellationToken = default". This is because you can call it like this GetCommentsStreamAsync(null).WithCancellation(new CancellationToken()) which then synchronizes it to the cancellation token in the method signature.

I'll get started now. 

Thanks!

**Comment by henrikfroehling** [comment]
Issue is resolved and PR is merged. 👍
